### PR TITLE
GIX-998: SNS transfer

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Add the method to make transfers in SNS ledger.

# Changes

* Add "transfer" method to ledger and wrapper classes.
* New TransferParams type used in "transfer" methods.
* New toTransferArg converters to transform TransferParams to TransferArg required by canister.

# Tests

* Test new method in ledger.
* Test new method in wrapper.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@mstrasinskis @peterpeterparker please review

File: packages/sns/src/converters/ledger.converteres.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { toNullable } from "@dfinity/utils";
```
Comment (peterpeterparker) [inline packages/sns/src/converters/ledger.converteres.ts:1]: typo in file name `ledger.converteres.ts`

File: packages/sns/src/errors/ledger.errors.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import type { TransferError } from "../../candid/icrc1_ledger";
+
+export class SnsTransferError extends Error {
+  public type: TransferError;
```
Comment (peterpeterparker) [inline packages/sns/src/errors/ledger.errors.ts:4]: `type` isn't a reserved keyword?

File: packages/sns/src/ledger.canister.ts [hunk 3]
```diff
@@ -40,4 +43,28 @@ export class SnsLedgerCanister extends Canister<SnsLedgerService> {
...
+      throw new SnsTransferError({
+        type: response.Err,
+        msg: "Failed to transfer",
+      });
+    }
+    if ("Ok" in response) {
+      return response.Ok;
+    }
+    // Edger case, should never happen.
+    throw new Error("Unexpected response from transfer");
```
Comment (peterpeterparker) [inline packages/sns/src/ledger.canister.ts:68]: isn't the `response` a variant? it can be something else than error or ok?

Comment (lmuntaner) [reply packages/sns/src/ledger.canister.ts:68]: It's a variant with `Ok` and `Error` only.

Comment (lmuntaner) [reply packages/sns/src/ledger.canister.ts:68]:
```
type TransferResult = variant {
    Ok : BlockIndex;
    Err : TransferError;
};
```

File: packages/sns/src/errors/ledger.errors.ts (hunk 2 above)
Comment (lmuntaner) [reply packages/sns/src/errors/ledger.errors.ts]: done

File: packages/sns/src/converters/ledger.converteres.ts (hunk 1 above)
Comment (lmuntaner) [reply packages/sns/src/converters/ledger.converteres.ts]: done, thanks!

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

File: packages/sns/src/errors/ledger.errors.ts (hunk 2 above)
Comment (peterpeterparker) [reply packages/sns/src/errors/ledger.errors.ts]: Oh good to know, thanks for double checking and feedback!

File: packages/sns/src/ledger.canister.ts (hunk 3 above)
Comment (peterpeterparker) [reply packages/sns/src/ledger.canister.ts:68]: Is there an edge case then, should we simplify it?

```
if ("Err" in response) {
      throw new SnsTransferError({
        type: response.Err,
        msg: "Failed to transfer",
      });
    }
    
return response.Ok;
```

In Motoko at least the Result type (https://internetcomputer.org/docs/current/references/motoko-ref/result/) is ok or err and cannot be extended I think.

Comment (lmuntaner) [reply packages/sns/src/ledger.canister.ts:68]: Sure

Comment (lmuntaner) [reply packages/sns/src/ledger.canister.ts]: done

**Comment by lmuntaner** [comment]
@peterpeterparker ready for another review

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks

Before merge you can add a mention in CHANGELOG about it or in a next PR

**Pull request merged**